### PR TITLE
add endpoint to bundle configuration

### PR DESCRIPTION
**Pull request by mohamedyasser**
## Subject
use digitalocean space with s3 sdk by allow changing endpoint

### screenshot
https://ta2h.ams3.cdn.digitaloceanspaces.com/default/0001/01/5723218d8e4b5ea01b64af37e09501a3525ad1ab.png
```markdown
### Added
- Added new `endpoint` configuration node under S3 filesystem configuration
```

### CODE REVIEW

File: src/DependencyInjection/SonataMediaExtension.php [hunk 1]
```diff
@@ -307,7 +307,7 @@ public function registerDoctrineMapping(array $config)
             'fieldName' => 'media',
             'targetEntity' => $config['class']['media'],
             'cascade' => [
-                 'persist',
```
Comment (core23) [inline src/DependencyInjection/SonataMediaExtension.php]: Please revert the formatting

**Comment by OskarStark** [comment]
Please add a testcase

File: src/DependencyInjection/SonataMediaExtension.php [hunk 2]
```diff
@@ -1,7 +1,5 @@
 <?php
-
```
Comment (greg0ire) [inline src/DependencyInjection/SonataMediaExtension.php]: Uuuuh… why did you remove all these blank lines?

File: src/DependencyInjection/SonataMediaExtension.php [hunk 3]
```diff
@@ -238,11 +199,10 @@ public function configureBuzz(ContainerBuilder $container, array $config)
...
         $container->getDefinition('sonata.media.buzz.browser')
             ->replaceArgument(0, new Reference($config['buzz']['connector']));
-
         foreach ([
-            'sonata.media.buzz.connector.curl',
-            'sonata.media.buzz.connector.file_get_contents',
-        ] as $connector) {
+                     'sonata.media.buzz.connector.curl',
+                     'sonata.media.buzz.connector.file_get_contents',
+                 ] as $connector) {
```
Comment (greg0ire) [inline src/DependencyInjection/SonataMediaExtension.php]: Why are you touching this?

File: src/DependencyInjection/SonataMediaExtension.php (hunk 2 above)
Comment (mohamedyasser) [reply src/DependencyInjection/SonataMediaExtension.php]: I sorry but phpstorm reformat code before commit, I'll revert to last changes

File: src/DependencyInjection/SonataMediaExtension.php (hunk 3 above)
Comment (greg0ire) [reply src/DependencyInjection/SonataMediaExtension.php]: And why did you mark this as resolved?

Comment (mohamedyasser) [reply src/DependencyInjection/SonataMediaExtension.php]: in the last commit, I'm reverting to last changes and add endpoint code only 
commit e76b3c56c6464323fc3118ca2e7976588b3a2df8

File: src/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -243,6 +243,7 @@ private function addFilesystemSection(ArrayNodeDefinition $node)
                                     ->end()
                                 ->end()
                                 ->scalarNode('region')->defaultValue('s3.amazonaws.com')->end()
+                                ->scalarNode('endpoint')->defaultNull()->end()
```
Comment (OskarStark) [inline src/DependencyInjection/Configuration.php]: please add `->info()` node with a meaningful description

Comment (OskarStark) [inline src/DependencyInjection/Configuration.php]: please add `->info()` node with a meaningful description

Comment (mohamedyasser) [reply src/DependencyInjection/Configuration.php]: if this good description or not

`->scalarNode('endpoint')->defaultNull()->info('custom bucket endpoint for any storage use s3 SDK like DigitalOcean Spaces')->end()`

**Comment by mohamedyasser** [comment]
@core23 @OskarStark any update

File: src/DependencyInjection/SonataMediaExtension.php (hunk 3 above)
Comment (greg0ire) [reply src/DependencyInjection/SonataMediaExtension.php]: Please revert that change too.

File: src/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -243,6 +243,7 @@ private function addFilesystemSection(ArrayNodeDefinition $node)
                                     ->end()
                                 ->end()
                                 ->scalarNode('region')->defaultValue('s3.amazonaws.com')->end()
+                                ->scalarNode('endpoint')->defaultNull()->info('custom bucket endpoint for any storage use s3 SDK like DigitalOcean Spaces')->end()
```
Comment (greg0ire) [inline src/DependencyInjection/Configuration.php]:
```suggestion
                                ->scalarNode('endpoint')
                                    ->defaultNull()
                                    ->info('custom bucket endpoint for any storage use s3 SDK like DigitalOcean Spaces')
                                ->end()
```

File: src/DependencyInjection/SonataMediaExtension.php (hunk 1 above)
Comment (greg0ire) [reply src/DependencyInjection/SonataMediaExtension.php]: I think this one can stay @core23 , it changes from 5 to 4 spaces

File: src/DependencyInjection/SonataMediaExtension.php [hunk 6]
```diff
@@ -345,9 +345,9 @@ public function registerDoctrineMapping(array $config)
                 'inversedBy' => null,
                 'joinColumns' => [
                     [
-                     'name' => 'category_id',
-                     'referencedColumnName' => 'id',
-                     'onDelete' => 'SET NULL',
+                        'name' => 'category_id',
+                        'referencedColumnName' => 'id',
+                        'onDelete' => 'SET NULL',
```
Comment (greg0ire) [inline src/DependencyInjection/SonataMediaExtension.php]: This one is correct too, and can stay IMO 👍

File: src/DependencyInjection/SonataMediaExtension.php [hunk 7]
```diff
@@ -451,11 +451,11 @@ public function configureFilesystemAdapter(ContainerBuilder $container, array $c
...
-                        'acl' => $config['filesystem']['s3']['acl'],
-                        'storage' => $config['filesystem']['s3']['storage'],
-                        'encryption' => $config['filesystem']['s3']['encryption'],
-                        'meta' => $config['filesystem']['s3']['meta'],
-                        'cache_control' => $config['filesystem']['s3']['cache_control'],
+                    'acl' => $config['filesystem']['s3']['acl'],
+                    'storage' => $config['filesystem']['s3']['storage'],
+                    'encryption' => $config['filesystem']['s3']['encryption'],
+                    'meta' => $config['filesystem']['s3']['meta'],
+                    'cache_control' => $config['filesystem']['s3']['cache_control'],
```
Comment (greg0ire) [inline src/DependencyInjection/SonataMediaExtension.php]: Same here 👍

File: src/DependencyInjection/SonataMediaExtension.php [hunk 8]
```diff
@@ -520,7 +524,7 @@ public function configureFilesystemAdapter(ContainerBuilder $container, array $c
             $container->getDefinition('sonata.media.adapter.filesystem.opencloud.connection')
                 ->replaceArgument(0, $config['filesystem'][$settings]['url'])
                 ->replaceArgument(1, $config['filesystem'][$settings]['secret'])
-                ;
+            ;
```
Comment (greg0ire) [inline src/DependencyInjection/SonataMediaExtension.php]: And here 👍

**Comment by greg0ire** [comment]
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already pushed. Don't forget the `--force` option otherwise git will try to merge both things together.

**Comment by greg0ire** [comment]
I edited your changelog to keep it purely functional

**Comment by greg0ire** [comment]
Blocked by #1540

File: .editorconfig [hunk 9]
```diff
@@ -7,7 +7,7 @@ charset = utf-8
 insert_final_newline = true
 trim_trailing_whitespace = true
 
-[*.{yml,twig,php}]
+[*.{yaml,yml,twig,php}]
```
Comment (greg0ire) [inline .editorconfig]: Not sure how you managed to obtain such a diff… I don't understand what is happening, I mean, 3.x is already like this.

Comment (greg0ire) [reply .editorconfig]: I think you have this weird history because you forgot to use `git push --force`

**Comment by greg0ire** [comment]
I just replaced it all with one clean commit.

**Comment by mohamedyasser** [comment]
> Can you add a test case with a configured endpoint value?

actual I can't do this because the access key and secret key it's private

**Comment by OskarStark** [comment]
> actual I can't do this because the access key and secret key it's private

I don't understand. You should write a testcase where a configured endpoint URL is really used in Adapter. You can ofc use mocks.

**Comment by mohamedyasser** [comment]
I don't understand. did you mean create mock test

**Comment by OskarStark** [comment]
Yes 👍🏻😊

File: src/DependencyInjection/SonataMediaExtension.php [hunk 10]
```diff
@@ -472,6 +472,10 @@ public function configureFilesystemAdapter(ContainerBuilder $container, array $c
                     ];
                 }
 
+                if (isset($config['filesystem']['s3']['endpoint'])) {
```
Comment (phansys) [inline src/DependencyInjection/SonataMediaExtension.php]: By default `$config['filesystem']['s3']['endpoint']` is always null, so we could save the call to `isset()`:
```suggestion
                if (null !== $config['filesystem']['s3']['endpoint']) {
```
